### ISSUE TITLE
Correctness of InfoS usage should be checked on whole codebase 

### ISSUE THREAD

**Issue by serathius** [kubernetes/kubernetes#102439]
For structured logging migration we have developed a tool to klog invocations [logcheck](https://github.com/kubernetes/klog/tree/master/hack/tools/logcheck)

For a given directory logcheck validates two things:
* directory is fully migrated - check if there are non-structured logging calls like `klog.Infof`
* usage of `klog.InfoS` and `klog.ErrorS` is correct - check if key values passed are paired.

We run logcheck as part of Kubernetes verify scripts [verify-strucured-logging.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/verify-structured-logging.sh) on directories listed in [.structured_logging](https://github.com/kubernetes/kubernetes/blob/master/hack/.structured_logging)

Directories listed in `.structured_logging` are marked as migrated to avoid regressions from organized migration, however there are also other code locations that already use new functions that are not yet ready to be marked as migrated (components that are not part of organized migrations are done as best effort).

#### What would you like to be added:

We should split checking if directory was migrated from correctness of new functions and scan whole codebase for correctness.

Plan:
* Add a flag to logcheck for requiring that directory was migrated.
* Add to verify-strucured-logging.sh running logcheck of whole k/k codebase to check correctness

#### Why is this needed:

This will ensure higher quality of migration PRs by validating correctness of directories that were not yet added in .structured_logging

/wg structured-logging
/cc @kubernetes/wg-structured-logging-members

**Comment by umangachapagain** [comment]
@serathius I'd like to pick this up. Do I need to wait till it is triage/accepted?

**Comment by serathius** [comment]
Feel free to pick it up :P

**Comment by umangachapagain** [comment]
> We should split checking if directory was migrated from correctness of new functions and scan whole codebase for correctness.
>
> Plan:
>
>   Add a flag to logcheck for requiring that directory was migrated.
>   Add to verify-strucured-logging.sh running logcheck of whole k/k codebase to check correctness
>
[**Case 1: Migration**] If the directory is marked as migrated, we check that it only uses correct format of `InfoS/ErrorS` and doesn't use `Infof/Errorf`.

[**Case 2: Correct usage**] If the directory is not marked as migrated, we check that it's usage of `InfoS/ErrorS` is correct and don't care about usage of `Infof/Errorf`.

By default, CI will test for correctness of whole k/k codebase instead of testing for complete migration (which is the  current behavior). And to test for migration, a new flag must be passed explicitly.

@serathius is this the expectation?

**Comment by serathius** [comment]
> By default, CI will test for correctness of whole k/k codebase instead of testing for complete migration (which is the current behavior). And to test for migration, a new flag must be passed explicitly.

To clarify, let me rephrase.

Current state is [Case 1: *Migration*] and [Case 2: *Correct usage*] tested on migrated directories from  [hack/.structured_logging](https://github.com/kubernetes/kubernetes/blob/master/hack/.structured_logging)

What we want:
* [Case 2: *Correct usage*] tested on whole repository
* [Case 1: *Migration*] tested on migrated directories from [hack/.structured_logging](https://github.com/kubernetes/kubernetes/blob/master/hack/.structured_logging)

**Comment by umangachapagain** [comment]
> What we want:
> 
>     * [Case 2: **Correct usage**] tested on whole repository
> 
>     * [Case 1: **Migration**] tested on migrated directories from [hack/.structured_logging](https://github.com/kubernetes/kubernetes/blob/master/hack/.structured_logging)
>
Ok. So this means we'll have to split the tests in logcheck into "correct usage" and "migration" tests. We can then run correct usage tests for entire codebase and migration tests for migrated directories.
If this is the case, I can start implementing it.

**Comment by serathius** [comment]
/milestone v1.22
Part of work go graduate Structured Logging KEP to beta.

**Comment by serathius** [comment]
I don't think work here has finished. We still need to implement those changes in k/k repo:
* Cut new klog
* Bump klog in k/k
* Update the script
/reopen

**Comment by umangachapagain** [comment]
> I don't think work here has finished. We still need to implement those changes in k/k repo:
> 
>     * Cut new klog
> 
>     * Bump klog in k/k
> 
>     * Update the script  ./hack/verify-structured-logging.sh
>
@serathius Thanks for reminding, I almost forgot 😓 

Why do we need to cut a new klog and bump in k/k? Can't we make latest logcheck available to k/k without it?
I don't know the process.

**Comment by serathius** [comment]
Hmm, We could try without updating to update. Please send PR that will update ./hack/verify-structured-logging.sh
If updating klog is nesesery feel free to contact @dims for help

**Comment by umangachapagain** [comment]
> Hmm, We could try without updating to update. Please send PR that will update ./hack/verify-structured-logging.sh
>
Should the script run logcheck on migrated packages only or entire codebase for correctness or both?

**Comment by serathius** [comment]
> If we do both, we'd have to run logcheck twice (with and without allow-unstructured flag) which means test time will increase.
This depends on implementation, increasing running time is ok as long it's reasonable, for example x1.5, but not x10.

**Comment by voigt** [comment]
Due to v1.22 code freeze, I'm moving this issue to the next release.

/milestone v1.23

**Comment by serathius** [comment]
We would like to have static analysis setup before we start migrating logs in 1.23. @umangachapagain will you be able to work on it this week? Do you need any help?

**Comment by umangachapagain** [comment]
> We would like to have static analysis setup before we start migrating logs in 1.23. @umangachapagain will you be able to work on it this week? Do you need any help?

@serathius Yes, I can. PR kubernetes/kubernetes/pull/103293 is up for enabling static analysis.  It has some unresolved conversations. It'd be helpful if I could get another round of review on that PR.

Before we merge that we'd need to fix existing logging issues which are listed here: https://github.com/kubernetes/kubernetes/pull/103293#issuecomment-885028839. If I could get help fixing those it'd be great. Else I can fix it too.

**Comment by serathius** [comment]
I see two issues that need to be addressed:
* Fix handling of `Errorf` in klog code, this should be done by creating an issue to klog
* Fix remaining errors, @umangachapagain  can you prepare a list of those errors and ask for help on #wg-structured-logging Slack channel?
